### ISSUE TITLE
DNS fails on image pull: dial tcp: lookup index.docker.io

### ISSUE THREAD

**Issue by AloisReitbauer** [kitematic/kitematic#718]
See error message below:

Get https://index.docker.io/v1/repositories/library/ubuntu/images: dial tcp: lookup index.docker.io on 192.168.90.1:53: read udp 192.168.90.1:53: i/o timeout

**Comment by mchiang0610** [comment]
This seems to be a problem with your DNS loading an older entry. 

Is this on Mac or Windows? 

Try: 

1. Click on the Docker CLI button 
![image](https://cloud.githubusercontent.com/assets/3325447/7950182/0ae55b3c-094c-11e5-859b-3acf43df7c34.png)

2. Type in `docker-machine ssh` 

3. `echo "nameserver 8.8.8.8" > /etc/resolv.conf`

**Comment by firrae** [comment]
Hi there,

I'm having the same issue and when I follow your answer I get the following `Error: Please specify a machine name.` After adding, what I assume is, the machine name I get the following: `Error: Host does not exist: hello-world.` I was trying to add the basic offical hello-world container to test it out.

Any further help is appreciated.

**Comment by rgbkrk** [comment]
If you run `docker-machine ls` it will show which you have available. That being said, mine is named `dev` (run `docker-machine ssh dev`).

**Comment by firrae** [comment]
Ah, thanks a bunch for that. Runs like a charm now.

**Comment by ambled** [comment]
I'm not sure if this is the/a cause or needs to be a separate issue:

my boot2docker.iso shows as:
Linux dev 4.0.7-boot2docker #1 SMP Wed Jul 15 00:01:41 UTC 2015 x86_64 GNU/Linux

I've not only had this DNS issue, but it keeps reoccurring every 12 hours.
I discovered that the dhcp configurations in this iso has some issues:

    * when the dhcp reservation (which might be low) expires, the /etc/resolv.conf gets rewritten by udhcpc
    * udhcpc uses  /usr/share/udhcpc/default.script
    * default.script either clobbers and rewrites only the dhcp response, or ...
    * you can override the DHCP list by creating a 'dns' environment variable entry in default.script as a whitelist seperated list of ips. However, default.script handles this case poorly and appends the configured nameservers every time the lease expires

the end result is that as configured (at least with the responses I get from my DHCP server) the VM might be working initially for people, then stops working because of a pause/resume machine event, network reconfiguration or just a DHCP lease timeout.

Since all of these events can trigger the problem and DHCP leases can vary widely between environments, it might be complicating efforts to find the root cause.

As this issue seems to come from upstream on the boot2docker.iso, my current workaround seems to be holding (bypassing the NAT DNS and forwarding to the host).

    docker-machine stop dev
    VBoxManage modifyvm dev --natdnsproxy1 off --natdnshostresolver1 off
    docker-machine start dev

There are lots of ways to address this, not sure which makes the most sense. I think I'd like the equivalent of the resovlconf framework with head and tail entries.  Improving the udhcpc script to look for /etc/resolvconf/resolvconf.d  or pull settings from /etc/default seems like a possible path, however that still ends up requiring manually fiddling files on the virtual host machine.

**Comment by ambled** [comment]
I suppose it's almost pointless to waste much time on the TinyCore image and revisit this subject after Debian is further along.

My current persistence hack, after starting the vm (but from OSX command line):

* Create docker machine boot patch script
    $ cat ~/.docker/bootsync.sh

        #!/bin/sh
        sed -i '/if \[ -n \"\$dns\" \]/a     dns=\"$dns 8.8.8.8\"' /usr/share/udhcpc/default.script
        #sed -i '/if \[ -n \"\$domain\" \]/a     domain=\"$domain dev\"' /usr/share/udhcpc/default.script

* Inject boot script into persistent location on virtual host image
    $ docker-machine scp ~/.docker/bootsync.sh dev:/mnt/sda1/var/lib/boot2docker/bootsync.sh

* Execute script to patch current ephemeral default.script. Or do a docker-machine restart
    $ docker-machine ssh dev sudo sh /mnt/sda1/var/lib/boot2docker/bootsynch.sh

**Comment by midwire** [comment]
Confirmed, once I overwrote the `dns-crypt` proxy settings in resolve.conf within the default VM with 'nameserver 8.8.8.8' it works.

**Comment by mchiang0610** [comment]
Is this still happening on the latest toolbox / Kitematic?

**Comment by mlushpenko** [comment]
Thanks @mchiang0610 , your suggestion helped me

**Comment by r4j4h** [comment]
@mchiang0610 your fix is still relevant as of 2/7/1026.

It just helped me on a fresh install of Docker toolbox v1.10.0. My resolv.conf had seemingly valid settings but they gave same results as initial issue reports using just straight docker-machine ssh default and then docker pull commands.

Setting to Google's DNS as you suggest instantly fixed the issue.

Here's what the .conf looked like before the fix, note the first line. I am wondering if it is the root cause:

```
search ph.cox.net
nameserver 68.105.**.**
```

**Comment by FrenchBen** [comment]
@r4j4h Unfortunately not all DNS are made equal - In your case it seems that your ISP's DNS isn't up to par. Using Google's DNS helps quite a bit (probably why they created the service in the first place)

**Comment by tylerhjones** [comment]
The problem is not the DNS you have, it is the packet return size. Im guessing for 8.8.8.8 the return size is less than 512 bytes. So the UDP packet is forwarded from the host mini vm running on your machine to the host machine. When the DNS returns a packet size over 512 DNS will fall over to TCP, which it looks like the vm is not forwarding to your host machine and thus it cannot resolve.

**Comment by FrenchBen** [comment]
@tylerhjones thanks for the detailed info. 
The `192.168` network is created by VirtualBox, which may not forward the TCP packet.

**Comment by timelf123** [comment]
>The problem is not the DNS you have, it is the packet return size. Im guessing for 8.8.8.8 the return size is less than 512 bytes. So the UDP packet is forwarded from the vm running on your machine to the host machine. When the DNS returns a packet size over 512 DNS will fall over to TCP, which it looks like the vm is not forwarding to your host machine and thus it cannot resolve.

This was it for me - `8.8.8.8 -> 208.67.222.222`  (opendns)

**Comment by Deepak-Poojari** [comment]
Your docker host may be behind the proxy. Set the proxy in /etc/sysconfig/docker. That should fix the problem.

**Comment by st0ne-dot-at** [comment]
Hi all,
i had similar issues with occasionally occuring lookup outages. I've tracked the issue down to the libc library of the gitea docker image.
The image uses the libc.musl-x86_64.so.1 libc library ... which uses an unusual resolve mechanism. It 
queries the nameservers in parallel (see [docker-alpine](http://gliderlabs.viewdocs.io/docker-alpine/caveats/)). This behavior differs from the glibc resolver behavior. The glibc resolver uses always the first entry.

In my case i had a local dns server in the first line, followed by external dns servers. This leaded to resolving errors, because only the local dns server was able to resolve the ip address of the db container.

May that helps.

**Comment by YongHaoWu** [comment]
@tylerhjones thanks for your explanation. 

>When the DNS returns a packet size over 512 DNS will fall over to TCP, which it looks like the vm is not forwarding to your host machine and thus it cannot resolve.

Why  `a packet size over 512` will fall over to TCP?

**Comment by WHK102** [comment]
Same problem from KVM image using CentOS7

```[root@server ~]# docker pull atlassian/jira-software
Using default tag: latest
Trying to pull repository docker.io/atlassian/jira-software ... 
Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 192.168.x.x:53: read udp 192.168.x.x:54328->192.168.x.x:53: i/o timeout```

why don't you use the default dns service of the system instead of guessing what the ip address is?
